app/main/tilt.py

```diff
@@ -0,0 +1,53 @@
+from datetime import datetime
+import time
+import bluetooth._bluetooth as bluez
+
+from . import blescan
+from .routes_tilt_api import process_tilt_data
+
+TILTS = {
+    'a495bb10c5b14b44b5121370f02d74de': 'Red',
+    'a495bb20c5b14b44b5121370f02d74de': 'Green',
+    'a495bb30c5b14b44b5121370f02d74de': 'Black',
+    'a495bb40c5b14b44b5121370f02d74de': 'Purple',
+    'a495bb50c5b14b44b5121370f02d74de': 'Orange',
+    'a495bb60c5b14b44b5121370f02d74de': 'Blue',
+    'a495bb70c5b14b44b5121370f02d74de': 'Yellow',
+    'a495bb80c5b14b44b5121370f02d74de': 'Pink',
+}
```

Comment:
There was just supported added to tiltpi to support more than one of each color.  I'm curious as to how they support that and if there is something similar we can do for that. I know at least 2 folks that have duplicate colors at this point that would benefit from that.

Reply:
mmh. will have to look into that. I am also not sure if anything differs for newer tilts in terms of the message format etc.

Anyway, really packing it in now. I am at a good stopping point. If a windows user with a tilt wants to try this test program to confirm it works on windows too, just `pip3 install bleak' and run this

```
import asyncio
import sys
import struct
from bleak import BleakScanner

IBEACON_PROXIMITY_TYPE = b"\x02\x15"

TILTS = {
    'a495bb10c5b14b44b5121370f02d74de': 'Red',
    'a495bb20c5b14b44b5121370f02d74de': 'Green',
    'a495bb30c5b14b44b5121370f02d74de': 'Black',
    'a495bb40c5b14b44b5121370f02d74de': 'Purple',
    'a495bb50c5b14b44b5121370f02d74de': 'Orange',
    'a495bb60c5b14b44b5121370f02d74de': 'Blue',
    'a495bb70c5b14b44b5121370f02d74de': 'Yellow',
    'a495bb80c5b14b44b5121370f02d74de': 'Pink',
    'a495ff10c5b14b44b5121370f02d74de': 'Black',
}

def get_number(data):
    integer = 0
    multiple = 256
    for c in data:
        integer += c * multiple
        multiple = 1
    return integer

def get_string(data):
    string = ''
    for c in data:
        string += '%02x' % c
    return string

async def run():
    print("Scanning for tilts...")
    devices = await BleakScanner.discover()
    for d in devices:
        #print(d)
        if d.metadata['manufacturer_data'] and 76 in d.metadata['manufacturer_data']:
            data = d.metadata['manufacturer_data'][76]
            #print(get_string(data))
            if bytearray(data[:2]) == bytearray(IBEACON_PROXIMITY_TYPE):
                uuid=get_string(data[2:18]).replace('-','')
                if uuid in TILTS:
                    color = TILTS[uuid]
                    temp=get_number(data[18:20]) # farenheight
                    gravity=get_number(data[20:22]) # gravity * 1000
                    print("uuid: {} color: {} temp: {} gravity: {}".format(uuid, color, temp, gravity))


loop = asyncio.get_event_loop()
loop.run_until_complete(run())
```

and when a tilt is found you'll see something like

```
Charless-MacBook-Pro:picobrew_pico cgalpin$ sudo python3 test_bleak.py 
Scanning for tilts...
uuid: a495bb30c5b14b44b5121370f02d74de color: Black temp: 67 gravity: 1010
```

Reply:
Guess really it wouldn't be hard to determine the difference between tilts cause each have a unique mac address. But the color is encoded in a UUID...

Let's make sure that any device-id in code are referring to the devices mac address and not the color code which doesn't uniquely identify a device, but rather simply identifies the color of the device.

Reply:
This would simply be `d.name`.
I'd like to also start tracking `rssi` information, but can add that later. I'm working on collecting this for the Z series, but testing has slowed cause I've disassembled everything.

Reply:
No, name is always 'Tilt' for me. There is a 'uuids' array but I am finding it is not always populated for some reason (but it differs from the uuids used to figure out the color). The address might be unique for all (although the format differs between mac and linux this shouldn't be a problem), but i am not sure since I only have one. If you could ask the people who have multiple tilts to run this version, it would help clear this up

Example output on mac
```
Charless-MacBook-Pro:picobrew_pico cgalpin$ sudo python3 test_bleak.py 
Scanning for tilts...
Found name: Tilt address: F0B45560-D29D-4717-B89A-2E6F9B9EA010 uuid: unknown metadata: {'uuids': [], 'manufacturer_data': {76: b'\x02\x15\xa4\x95\xbb0\xc5\xb1KD\xb5\x12\x13p\xf0-t\xde\x00C\x03\xf2\xc5'}, 'delegate': <CentralManagerDelegate: 0x7fe95f43e460>}
uuid: a495bb30c5b14b44b5121370f02d74de color: Black temp: 67 gravity: 1010 rssi: 50432
```

Example output on linux
```
pi@raspberrypi:~/picobrew_pico $ python3 test_bleak.py 
Scanning for tilts...
Found name: Tilt address: 50:65:83:6D:9F:70 uuid: a495ff10-c5b1-4b44-b512-1370f02d74de metadata: {'uuids': ['a495ff10-c5b1-4b44-b512-1370f02d74de'], 'manufacturer_data': {76: [2, 21, 164, 149, 187, 48, 197, 177, 75, 68, 181, 18, 19, 112, 240, 45, 116, 222, 0, 67, 3, 242, 197]}}
uuid: a495bb30c5b14b44b5121370f02d74de color: Black temp: 67 gravity: 1010 rssi: 50432
```

I am finding the rssi is always the same - not sure if that will be useful for the tilt. I am pretty sure I am getting it from the right byte.

Script
```
import asyncio
from bleak import BleakScanner

IBEACON_PROXIMITY_TYPE = b"\x02\x15"

TILTS = {
    'a495bb10c5b14b44b5121370f02d74de': 'Red',
    'a495bb20c5b14b44b5121370f02d74de': 'Green',
    'a495bb30c5b14b44b5121370f02d74de': 'Black',
    'a495bb40c5b14b44b5121370f02d74de': 'Purple',
    'a495bb50c5b14b44b5121370f02d74de': 'Orange',
    'a495bb60c5b14b44b5121370f02d74de': 'Blue',
    'a495bb70c5b14b44b5121370f02d74de': 'Yellow',
    'a495bb80c5b14b44b5121370f02d74de': 'Pink',
    'a495ff10c5b14b44b5121370f02d74de': 'Black',
}

def get_number(data):
    integer = 0
    multiple = 256
    for c in data:
        integer += c * multiple
        multiple = 1
    return integer

def get_string(data):
    string = ''
    for c in data:
        string += '%02x' % c
    return string

async def run():
    print("Scanning for tilts...")
    devices = await BleakScanner.discover()
    for d in devices:
        #print(d)
        if d.metadata['manufacturer_data'] and 76 in d.metadata['manufacturer_data']:
            data = d.metadata['manufacturer_data'][76]
            #print(get_string(data))
            if bytearray(data[:2]) == bytearray(IBEACON_PROXIMITY_TYPE):
                uuid=get_string(data[2:18]).replace('-','')
                if uuid in TILTS:
                    #### begin test theory on real uuid (reliably being address)
                    real_uuid = d.metadata['uuids'][0] if ('uuids' in d.metadata and len(d.metadata['uuids']) > 0) else "unknown"
                    print("Found name: {} address: {} uuid: {} metadata: {}".format(d.name, d.address, real_uuid, d.metadata))
                    #### end test theory
                    color = TILTS[uuid]
                    temp=get_number(data[18:20]) # farenheight
                    gravity=get_number(data[20:22]) # gravity * 1000
                    rssi=get_number(data[-1:])
                    print("uuid: {} color: {} temp: {} gravity: {} rssi: {}".format(uuid, color, temp, gravity, rssi))


loop = asyncio.get_event_loop()
loop.run_until_complete(run())
```

Reply:
I know that rssi changes, in the tiltpi code it tacks the minimum rssi. Bluetooth at close ranges typically has a high signal strength indication especially when there aren't obstructions (doors, metal, etc) if you go to a longer range you should see this reflected in the received signal strength indicated (RSSI) value.

I don't have duplicates, but I know enough of the tiltpi code as I've contributed some local changes to it and some have been merged into the master flow.json file (man do I hate `node-RED`). I'll check this out and see what help I can bring in identifying this value. The MAC address is the unique identity of any Bluetooth signal same as with WiFi it is the unique device it isn't in the beacon payload, but rather part of the protocol header.

Here is the data when I scan for any beacons using  `sudo python3 -u -m aioblescan -T`

```
{"uuid": "a495bb60c5b14b44b5121370f02d74de", "major": 46, "minor": 1012, "tx_power": -59, "rssi": -72, "mac": "c7:f9:f4:a8:65:1e"}
{"uuid": "a495bb60c5b14b44b5121370f02d74de", "major": 46, "minor": 1012, "tx_power": -59, "rssi": -69, "mac": "c7:f9:f4:a8:65:1e"}
{"uuid": "a495bb60c5b14b44b5121370f02d74de", "major": 46, "minor": 1012, "tx_power": -59, "rssi": -72, "mac": "c7:f9:f4:a8:65:1e"}
{"uuid": "a495bb60c5b14b44b5121370f02d74de", "major": 46, "minor": 1012, "tx_power": -59, "rssi": -73, "mac": "c7:f9:f4:a8:65:1e"}
{"uuid": "a495bb60c5b14b44b5121370f02d74de", "major": 46, "minor": 1012, "tx_power": 9, "rssi": -72, "mac": "c7:f9:f4:a8:65:1e"}
```

Just a few feet away from the catalyst that has 5g of lager is showing up with a few different rssi values and this `mac` is the unique identity of the device whereas `uuid` (awful name for this value... should be `colorUuid` or something) is identifying BLUE.

Reply:
I'm now curious as to why macos sees a different address value there than say on the raspberrypi I'd expect them to see the same unique identity/address of the device.

Reply:
Guess that is the main reason one should read docs... lol this is expected.

> The first part, **_a Bluetooth address in Windows and Linux and a UUID in macOS_**, is what is used for connecting to a device using Bleak. The list of objects returned by the discover method are instances of bleak.backends.device.BLEDevice and has name, address and rssi attributes, as well as a metadata attribute, a dict with keys uuids and manufacturer_data which potentially contains a list of all service UUIDs on the device and a binary string of data from the manufacturer of the device respectively.

So yeah let's use that identifier as the local `uid` / `device id` or whatever else we call it. These will uniquely identify the device. On my MacBook (which is too far away from my tilt) gets a bunch of these UUIDs for various bluetooth devices within range.

```
python3 scan_tilt.py
Scanning for tilts...
16A82A25-1F24-4868-BBB0-8D59F9A7C7C3: Apple, Inc. (b'\x0f\x08\xc0\nZ\x98\xae\x00@\x0b\x10\x02\x01\x04')
FEDEC37D-25A9-415C-A926-88ADB0ACAB8C: Unknown
B4617E01-8333-4931-A59B-7E4DAAD90EF9: Nest Cam
EE4FBD83-AD37-4A33-8023-05E344293EBB: Sony Corporation (b'\x02\x00\x01\x10\xae^\xba.r\xf2K\xe3\x88q|>\xe9\xdd \xe4\xacG>8\x0eB')
11CD1581-CEAD-47E7-A36A-ACBAE66F662C: Apple, Inc. (b'\x0f\x08\xc0\nq?\x07\x00@\x0b\x10\x02\x01\x04')
29FAA156-218A-4B7A-B913-0A074EAE7ACA: Apple, Inc. (b'\t\x06\x02\x00\x00\x00\x00\x00')
28268C95-C7D6-4F81-8F3E-814BEEFDB95B: Samsung Electronics Co. Ltd. (b'B\x04\x01\x80`d\x1c\xb0\xad\x99\xeff\x1c\xb0\xad\x99\xee\x01\x00\x00\x00\x00\x00\x00')
DC920CF4-D21E-43AB-9F59-B9C28AC4AC99: Google (b'V\x17\xca\x9b\xe9\xc3')
BEA6E203-5BD7-44C9-8693-6703B26DE931: Hewlett-Packard Company (b'\x01\xc9\x06')
63938092-D9C8-48DD-8D01-A0A052EE63C3: Apple, Inc. (b'\t\x06\x03\x07\xc0\xa8\xa8e')
4E77ABD1-8924-42F5-8BC8-4FD6A55B061D: Google (b'V\xea\xca<\xfa\x9c')
A53B512E-6A33-4CFB-8EF8-63EA7527D7C7: Unknown
952B599C-FB11-4282-91BF-D0E2C9460AF3: Unknown
64E44932-8EE2-4A37-AD24-D39BDD10C10B: Google (b'V\x18\xcav!\xd6')
C1E0B495-685F-48DF-B833-7DC7B9296A4B: Apple, Inc. (b'\x10\x07$\x1fH`4\xaa\x18')
D15CC511-18FE-48B0-8218-0C4BA371BA0A: Samsung Electronics Co. Ltd. (b'B\x04\x01\x80\xa0\xbc\x14\x85f\x05N&K\x03po\x9d\x01\x00\x00\x00\x00\x00\x00')
DB266B02-197C-4368-A8C7-B3D912C0EFC3: Unknown
A450C193-6372-4592-863D-8DE9388A3670: Apple, Inc. (b'\x10\x05J\x1c\x97J3')
802C6550-4573-4281-A871-4E4F7DBEEA71: Samsung Electronics Co. Ltd. (b'B\x04\x01\x80fD\\\xe9\xdb.\xddF\\\xe9\xdb.\xdc\x01\x00\x00\x00\x00\x00\x00')
C637AB51-D349-4C5E-A62F-F826953967A5: Unknown
1D070EB5-D7AE-4B9A-90E6-CE813557787A: Unknown
208CEE3F-25E4-43E1-BC70-C312E80AD25F: Google (b'V#\xca0T\x84')
```

Reply:
yup, mac is the same as address in bleak. On linux it looks like yours, and on a mac it looks like a GUID. What i did was use the uuid from the metadata if it's there (linux) otherwise use the address/guid (mac). I think this will prove to be what we need to support multiple tilts.

I must not get getting the right byte for RSSI, but i committed the changes to use bleak instead of pybluez. I'll figure ou the rssi this weekend.

Reply:
Oh i did not see your last remark. I chose to use the metadata uuid because it looked unique (different from the colors) but if that proves not to be the case we can use the bluetooth address and strip ':' or convert to '-'s

BTW, we really need to make it easy to rename a device from the main screen. Or list devices without aliases on the devices page as well. Right now it's a pita to add the alias :)

Reply:
You can rename devices from the device screen, not from all the brew graphs.

PM me on FB if you need help figuring that out.

Hrm not sure what that UUID is, but sure we can start there and if run into problem we can always change.

Reply:
Though to be consistent with tiltpi it would be the bluetooth address identifier which is different between Mac and Linux/Windows as described above.

Not sure what this "beacon service uuid" is nor if it is consistently unique as it is provided by the manufacturer (Baronbrew) and not from the bluetooth transmitter hardware (like address would be). Since Baronbrew built and support tiltpi I'd imagine they chose Mac Address for a reason 🤷‍♂️.

Reply:
yeah, i was about to make another merge, because i left it running for a while, and found the uuid shows up on the mac occasionally which gives phantom devices. Will use address aka mac.

Reply:
using mac address for uuid. multiple tilts of the same color should be supported.